mirascope/openai/tool_streams.py

```diff
@@ -0,0 +1,216 @@
+                    current_tool_call,
+                    current_tool_type,
+                    False,
+                )
+        return None, current_tool_call, current_tool_type, False
+
+    if not chunk.tool_calls:
+        return None, current_tool_call, current_tool_type, False
+
+    tool_call = chunk.tool_calls[0]
+    # Reset on new tool
+    if tool_call.id and tool_call.function is not None:
+        previous_tool_call = current_tool_call.model_copy()
+        previous_tool_type = current_tool_type
+        current_tool_call = ChatCompletionMessageToolCall(
+            id=tool_call.id,
+            function=Function(
+                arguments="",
+                name=tool_call.function.name if tool_call.function.name else "",
+            ),
+            type="function",
+        )
+        current_tool_type = None
+        for tool_type in chunk.tool_types:
+            if tool_type.__name__ == tool_call.function.name:
+                current_tool_type = tool_type
+                break
+        if current_tool_type is None:
+            raise RuntimeError(
+                f"Unknown tool type in stream: {tool_call.function.name}"
+            )
+        if previous_tool_call.id and previous_tool_type is not None:
+            return (
+                previous_tool_type.from_tool_call(
+                    previous_tool_call, allow_partial=allow_partial
+                ),
+                current_tool_call,
+                current_tool_type,
+                True,
+            )
+
+    # Update arguments with each chunk
+    if tool_call.function and tool_call.function.arguments:
+        current_tool_call.function.arguments += tool_call.function.arguments
+
+        if allow_partial and current_tool_type:
+            return (
+                partial(current_tool_type).from_tool_call(
+                    current_tool_call, allow_partial=True
+                ),
+                current_tool_call,
+                current_tool_type,
+                False,
+            )
+
+    return None, current_tool_call, current_tool_type, False
+
+
+class OpenAIToolStream(BaseToolStream[OpenAICallResponseChunk, OpenAITool]):
+    """A base class for streaming tools from response chunks."""
+
+    @classmethod
+    @overload
+    def from_stream(
+        cls,
+        stream: Generator[OpenAICallResponseChunk, None, None],
+        allow_partial: Literal[True],
+    ) -> Generator[Optional[OpenAITool], None, None]:
+        yield ...  # type: ignore  # pragma: no cover
+
+    @classmethod
+    @overload
+    def from_stream(
+        cls,
+        stream: Generator[OpenAICallResponseChunk, None, None],
+        allow_partial: Literal[False],
+    ) -> Generator[OpenAITool, None, None]:
+        yield ...  # type: ignore  # pragma: no cover
+
+    @classmethod
+    @overload
+    def from_stream(
+        cls,
+        stream: Generator[OpenAICallResponseChunk, None, None],
+        allow_partial: bool = False,
```

Comment:
why do you need this if you cover literal true and false above

Reply:
because those don't cover `bool`... thanks mypy